### ISSUE TITLE
Should value property accept null and undefined values to clear the input?

### ISSUE THREAD

**Issue by AriasBros** [uNmAnNeR/imaskjs#173]
Hi, currently, if we try set the value of the mask to null (or undefined), we will get the next exception: `TypeError: Cannot read property 'length' of null`

Example:

```
describe('#update value', function () {
  it('should update mask and set the value to empty string when value is set to null', function () {
    imask.updateOptions({
      mask: Number
    });

    const value = "100";
    imask.value = value;
    assert.strictEqual(imask.value, value);

    imask.value = null;
    assert.strictEqual(imask.value, "");
  });
});
```

Here is the code trace that we can get running the previous test:

```
  1) InputMask
       #update value
         should update mask and set the value to empty string when value is set to null:
     TypeError: Cannot read property 'length' of null
      at MaskedNumber.length (src/masked/base.js:254:27)
      at MaskedNumber.append [as resolve] (src/masked/base.js:122:10)
      at MaskedNumber.resolve [as value] (src/masked/base.js:116:10)
      at InputMask.set value [as value] (src/controls/input.js:96:5)
      at Context.<anonymous> (test/controls/input.js:113:7)
```

If this is not the desired behavior I can try send a PR where, when we set value to null/undefined, value is reset to empty string.

**Comment by uNmAnNeR** [comment]
Hi! Currently it is working as supposed to be. But we can rethink this point. What I had in mind when was going this way is to be able to maintain it easily leaving all preparations to user. But I think this is about convention and community. Maybe you can say the common way of handling null and undefined in such cases in js world? Maybe some examples in popular libraries?
So I don't mind to handle nulls, just want to be sure and have some base to reason about.

**Comment by AriasBros** [comment]
Hello! IMO I think in this case the better solution could be mimic the input HTML tag behavior:

https://jsfiddle.net/8hsawde2/11/

As we can see, the input element accepts `null` values to clear the input.
On the other hand, using `undefined` values are parsed to string and the input displays "undefined", just as it is.

So, because display "undefined" in the input does not have sense in a masked input, we could unify both cases in the `null` case.

> What I had in mind when was going this way is to be able to maintain it easily leaving all preparations to user

I can understand this and it is not a bad design absolutely, but...

...I think leave the exceptions to be thrown without any feedback is pernicious for the developer, because if this behavior is not documented you are forcing to the developer to debug the library. For example this case. The user the only feedback that is receiving is this exception:

    TypeError: Cannot read property 'length' of null
      at MaskedNumber.length (src/masked/base.js:254:27)
      at MaskedNumber.append [as resolve] (src/masked/base.js:122:10)
      at MaskedNumber.resolve [as value] (src/masked/base.js:116:10)
      at InputMask.set value [as value] (src/controls/input.js:96:5)
      at Context.<anonymous> (test/controls/input.js:113:7)

And it is not obvious that the error is happening when you are setting the masked value to `null`.

> Maybe you can say the common way of handling null and undefined in such cases in js world? Maybe some examples in popular libraries?

A good practice is to use assertions in points where you think they can be error prone. The assertions are displaying in development environments and removed in production ones. Using this, you are supplying more context that will help to the developer debug any exception that they could have.

Example from Ember.js:

https://github.com/emberjs/ember.js/blob/master/packages/@ember/-internals/runtime/lib/mixins/observable.js#L437

An example applied to this case could be:

    set(value) {
        assert('Value can't be null or undefined', value !== null && value !== undefined);
        or 
        assert('Value must be a valid String value', isString(value));
        or whatever
        ...
    }

In this way, developer will see a message error in console `Value can't be null or undefined` with the exception trace.

**Comment by uNmAnNeR** [comment]
Thanks for detailed explanation and example from `emberjs`!

> IMO I think in this case the better solution could be mimic the input HTML tag behavior

But it's still not fully consistent with HTML input, because HTML tag also casts other types to string (like Dates, Numbers). I think this is almost always not desired behaviour, especially in case of mask. So could only nulls handling really help?
I also tend to keep strict API for the reason of simplicity. I've just imagined how to implement nulls processing and realized that it should be done in many places, because internal `append` method (used in many other places) still should handle string strictly. It just does not make sense with null. Therefore nulls handling should be done in that places before calling `append`.

> I think leave the exceptions to be thrown without any feedback is pernicious for the developer

Definitely this is annoying thing, but this is the cost for using dynamic types. I supposed assertions are not designed as replacement of type system. I got such errors very often, and for me this is huge disadvantage of using dynamicly typed language. But still it seems community is agreed with that point when choosing javascript. IMO typescript or flow can help better than assertions. Let me think for a while to consider using assertions in some critical places with https://github.com/unassert-js/babel-plugin-unassert.
